mapper/tedge_mapper/src/mapper.rs

```diff
@@ -86,26 +91,103 @@ impl Mapper {
 
     #[instrument(skip(self), name = "messages")]
     async fn subscribe_messages(&self) -> Result<(), MqttClientError> {
-        let mut messages = self.client.subscribe(self.config.in_topic.filter()).await?;
+        let mut messages = self
+            .client
+            .subscribe(self.config.in_topic_filter.clone())
+            .await?;
+
+        let mut children: HashSet<String> = HashSet::new();
+
         while let Some(message) = messages.next().await {
             debug!("Mapping {:?}", message.payload_str());
-            match self.converter.convert(message.payload_str()?) {
-                Ok(mapped) => {
-                    self.client
-                        .publish(mqtt_client::Message::new(&self.config.out_topic, mapped))
-                        .await?;
-                }
-                Err(error) => {
-                    debug!("Mapping error: {}", error);
+            debug!("Topic {:?}", message.topic.name);
+
+            // ? Validation for child device ID?
+            match get_child_id_from_topic(message.clone().topic.name) {
+                Some(child_id) if child_id.is_empty() => {
+                    let error = MapperError::InvalidChildId { id: child_id };
+                    error!("Child ID error: {}", error);
                     self.client
                         .publish(mqtt_client::Message::new(
                             &self.config.errors_topic,
                             error.to_string(),
                         ))
                         .await?;
                 }
+                Some(child_id) => {
+                    info!("Child ID {:?}", child_id);
+                    dbg!(&children);
+                    if !children.contains(&child_id) {
+                        children.insert(child_id.clone());
+                        if let Some(message) = self
+                            .converter
+                            .convert_child_device_creation(child_id.as_str())
+                        {
+                            self.client.publish(message).await?;
+                        };
+                    }
+
+                    match self
+                        .converter
+                        .convert_child_device_payload(message.payload_str()?, child_id.as_str())
+                    {
+                        Ok(mapped) => {
+                            self.client
+                                .publish(mqtt_client::Message::new(&self.config.out_topic, mapped))
```

Comment:
Since this `publish` and the `publish` for child device creation message in line no.126 are going to two different topics which don't guarantee any serialisation, isn't there the possibility of a race here where the measurement might reach the cloud first before the device creation message itself?